### PR TITLE
Feature/add support for metadata filtering

### PR DESCRIPTION
**Pull request by avaxaleph**
# Description
The DocumentIndex API now supports metadata filtering through the new `filter-index` feature. Based on this, the `DocumentIndex` client in the Intelligence-Layer was updated with the following:
- ability to create a filter-index in a namespace
- ability to assign/unassign a filter-index to/from an existing search index
- ability to list assigned filter-indexes for a search index
- ability to search in a collection with metadata filtering by following the allowed criteria and data types allowed by the DI API

Additionally, I've updated the tutorial notebook for the `DocumentIndex` to reflect how to use this functionality.

## Before Merging
 - [x] Review the code changes
    - Unused print / comments / TODOs
    - Missing docstrings for functions that should have them
    - Consistent variable names
    - ...
 - [x] Update `changelog.md` if necessary
 - [ ] Commit messages should contain a semantic [label](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) and the ticket number
   - Consider squashing if this is not the case

### CODE REVIEW

**Comment by SebastianNiehusAA** [comment]
Thank you for the PR. Really cool feature. I don't see any issues with the code and it's also very useful that you added the section to the notebook (I just fixed some typos and formatting issues there).
@filippoberga there are two things we should discuss before merging:
 - Adding tests for the filtering methods
 - Adding a small how-to notebook

**Review by SebastianNiehusAA — CHANGES_REQUESTED** [review]
TODO for IL Team
 - Tests
 - How-to notebook
 - Add CHANGELOG entry

**Comment by avaxaleph** [comment]
> Thank you for the PR. Really cool feature. I don't see any issues with the code and it's also very useful that you added the section to the notebook (I just fixed some typos and formatting issues there). @filippoberga there are two things we should discuss before merging:
> 
> * Adding tests for the filtering methods
> * Adding a small how-to notebook

Thanks @SebastianNiehusAA  ! I've also wanted to add the tests but I saw that a certain DI collection is used there and I don't have access to it.

**Comment by SebastianNiehusAA** [comment]
> > Thank you for the PR. Really cool feature. I don't see any issues with the code and it's also very useful that you added the section to the notebook (I just fixed some typos and formatting issues there). @filippoberga there are two things we should discuss before merging:
> > 
> > * Adding tests for the filtering methods
> > * Adding a small how-to notebook
> 
> Thanks @SebastianNiehusAA ! I've also wanted to add the tests but I saw that a certain DI collection is used there and I don't have access to it. The token is injected during the CI build, right?

We inject the necessary tokens during CI, yes. For me it is also working locally. Normally you only need the "usual" tokens as for the other tests in IL, e.g. the AA_TOKEN. Have you set up your tokens as outlined in env.sample?

**Comment by avaxaleph** [comment]
@SebastianNiehusAA @filippoberga I've added the test. You can review the PR again.

File: src/documentation/document_index.ipynb [hunk 1]
```diff
@@ -291,6 +291,223 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# 3\n",
+    "# assign our new filter-index to our collection\n",
+    "document_index.assign_filter_index_to_search_index(\n",
+    "    collection_path=collection_path,\n",
+    "    index_name=\"asym-128\",  # we assign it to asym-128\n",
```
Comment (SebastianNiehusAA) [inline src/documentation/document_index.ipynb:376]: If we are using `intelligence-layer-sdk-demo-collection` as a collection as assigned in the third cell, there does not seem to be an `asym-128` index. We could use `intelligence-layer-sdk-demo-index` instead.

File: src/documentation/document_index.ipynb [hunk 2]
```diff
@@ -291,6 +291,223 @@
...
+    "    filter_index_name=\"demo-title\",\n",
+    ")\n",
+    "\n",
+    "# check if our filter-index is assigned to our collection\n",
+    "document_index.list_assigned_filter_index_names(\n",
+    "    collection_path=collection_path, index_name=\"asym-128\"\n",
```
Comment (SebastianNiehusAA) [inline src/documentation/document_index.ipynb:382]: see comment above (asym-128 does not exists in the current collection)

File: src/documentation/document_index.ipynb [hunk 3]
```diff
@@ -291,6 +291,223 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "document_index_retriever = DocumentIndexRetriever(\n",
+    "    document_index=document_index,\n",
+    "    index_name=\"asym-128\",\n",
```
Comment (SebastianNiehusAA) [inline src/documentation/document_index.ipynb:401]: see comment above (asym-128 does not exists in the current collection)

**Review by SebastianNiehusAA — CHANGES_REQUESTED** [review]
Tests are looking good. I only found an issue with the index name that needs fixing

**Pull request merged**